src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+     *
+     * @var int[]
+     */
+    protected $index2word = [];
+
+    /**
+     * An array containing the hidden layer.
+     *
+     * @var Vector[]
+     */
+    protected $syn1 = [];
+
+    /**
+     * An array containing word vectors.
+     *
+     * @var Vector[]
+     */
+    protected $vectors = [];
+
+    /**
+     * The error vector used during training.
+     *
+     * @var \Tensor\Vector
+     */
+    protected $error;
+
+    /**
+     * The lock factors for each word in the context.
+     *
+     * @var int[]
+     */
+    protected $vectorsLockf = [];
+
+    /**
+     * The L2-normalized word vectors from the model.
+     *
+     * @var Vector[]
+     */
+    protected $vectorsNorm = [];
+
+    /**
+     * The cumulative distrubtion table for negative sampling.
+     *
+     * @var int[]
+     */
+    protected $cumTable = [];
+
+    /**
+     * The last digit in the cumulative distribution table.
+     *
+     * @var int
+     */
+    protected $endCumDigit;
+
+    /**
+     * The negative labels used in the cumulative distrubtion table.
+     *
+     * @var \Tensor\Vector
+     */
+    protected $negLabels;
+
+    /**
+     * The training method determined by the layer they determine.
+     *
+     * @var string
+     */
+    protected $trainMethod;
+
+    /**
+     * The layer of the network, accepts 'neg' or 'hs'.
+     *
+     * @var string
+     */
+    protected $layer;
+
+    /**
+     * The window size for the skip-gram model.
+     *
+     * @var int
+     */
+    protected $window;
+
+    /**
+     * Vector dimension size.
+     *
+     * @var int
+     */
+    protected $dimensions;
+
+    /**
+     * The degree to which noise words are removed from the training text.
+     *
+     * @var float
+     */
+    protected $sampleRate;
+
+    /**
+     * The amount of L2 regularization applied to the weights of the output layer.
+     *
+     * @var float
+     */
+    protected $alpha;
+
+    /**
+     * The maximum number of training epochs. i.e. the number of times to iterate
+     * over the entire training set before terminating.
+     *
+     * @var int
+     */
+    protected $epochs;
+
+    /**
+     * The minimum times a word must appear in the corpus to be considered in the training text.
+     *
+     * @var int
+     */
+    protected $minCount;
+
+    /**
+     * The total number of unique words in the corpus.
+     *
+     * @var int
+     */
+    protected $vocabCount;
+
+    /**
+     * Sigmoid activation function.
+     *
+     * @var \Rubix\ML\NeuralNet\ActivationFunctions\Sigmoid
+     */
+    protected $sigmoid;
+
+    /**
+     * @param string $layer
+     * @param int $window
+     * @param int $dimensions
+     * @param float $sampleRate
+     * @param float $alpha
+     * @param int $epochs
+     * @param int $minCount
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(
+        string $layer = 'neg',
+        int $window = 2,
+        int $dimensions = 5,
```

Comment:
I think this should be the first argument to the constructor as it depends highly on factors that vary between use-cases such as the size of the user's corpus. It would also be consistent with other Transformers that do embedding/dimensionality reduction.

Reply:
You want the dimensions to be the first hyper-parameter? There are 3 in the snippet, just making sure :)